src/media/local-stream.ts

```diff
@@ -141,55 +144,80 @@ abstract class _LocalStream extends Stream {
+    }
 
-    // Check that the loaded effect is the latest one and dispose if not
-    if (effect !== this.loadingEffects.get(name)) {
+    // Load the effect. Because loading is asynchronous, keep track of the loading effects.
+    this.loadingEffects.set(effect.kind, effect);
+    await effect.load(this.outputTrack);
+
+    // After loading, check whether or not we still want to use this effect. If another effect of
+    // the same kind was added while this effect was loading, we only want to use the latest effect,
+    // so dispose this one. If the effects list was cleared while this effect was loading, also
+    // dispose it.
+    if (effect !== this.loadingEffects.get(effect.kind)) {
       await effect.dispose();
-      throw new Error(`Effect "${name}" not required after loading`);
+      throw new WebrtcCoreError(
+        WebrtcCoreErrorType.ADD_EFFECT_FAILED,
+        `Another effect with kind ${effect.kind} was added while effect ${effect.id} was loading, or the effects list was cleared.`
+      );
     }
+    this.loadingEffects.delete(effect.kind);
 
-    // Use the effect
-    this.loadingEffects.delete(name);
-    this.effects.push({ name, effect });
-    this.changeOutputTrack(outputTrack);
+    // Add the effect to the effects list.
+    this.effects.push(effect);
 
     // When the effect's track is updated, update the next effect or output stream.
     // TODO: using EffectEvent.TrackUpdated will cause the entire web-media-effects lib to be built
     // and makes the size of the webrtc-core build much larger, so we use type assertion here as a
     // temporary workaround.
     effect.on('track-updated' as EffectEvent, (track: MediaStreamTrack) => {
-      const effectIndex = this.effects.findIndex((e) => e.name === name);
+      const effectIndex = this.effects.indexOf(effect);
       if (effectIndex === this.effects.length - 1) {
         this.changeOutputTrack(track);
       } else {
-        this.effects[effectIndex + 1]?.effect.replaceInputTrack(track);
+        this.effects[effectIndex + 1]?.replaceInputTrack(track);
       }
     });
+
+    // Emit an event with the effect so others can listen to the effect events.
+    this[LocalStreamEventNames.EffectAdded].emit(effect);
   }
 
   /**
-   * Get an effect from the effects list.
+   * Get an effect from the effects list by ID.
    *
-   * @param name - The name of the effect you want to get.
+   * @param id - The id of the effect you want to get.
    * @returns The effect or undefined.
    */
-  getEffect(name: string): TrackEffect | undefined {
-    return this.effects.find((e) => e.name === name)?.effect;
+  getEffectById(id: string): TrackEffect | undefined {
+    return this.effects.find((effect) => effect.id === id);
+  }
+
+  /**
+   * Get all the effects from the effects list with the given kind.
+   *
+   * @param kind - The kind of the effects you want to get.
+   * @returns A list of effects.
+   */
+  getEffectsByKind(kind: string): TrackEffect[] {
```

Comment:
This returning an array seems inconsistent with the fact that we only allow one instance of each effect type

Reply:
Sure, will change it.

Reply:
just curious , if we knew there is only 3 kind of effects why didn't we have some function like addVirtualBackgroundEffect rather have names passed in to the effects ?

Reply:
Well there are 3 effects _now_, but we want to support any kind of effect in the future as opposed to hard-coding the current ones in.